osquery/tables/system/linux/services.cpp

```diff
@@ -0,0 +1,774 @@
+
+  return enabled.get();
+}
+
+bool upstartEnabled() {
+  static boost::optional<bool> enabled;
+  if (enabled != boost::none) {
+    return enabled.get();
+  }
+
+  static std::mutex m;
+  std::lock_guard<std::mutex> lock(m);
+  if (enabled != boost::none) {
+    return enabled.get();
+  }
+
+  enabled = boostfs::exists("/etc/init");
+  VLOG(1) << "Upstart was" << (enabled.get() ? " " : " NOT ") << "found";
+
+  return enabled.get();
+}
+
+Status enumerateSysVinitServices(QueryData& query_data) {
+  if (!sysVinitEnabled()) {
+    return Status(0, "OK");
+  }
+
+  static const std::vector<std::string> run_level_dir_list = {"/etc/rc0.d",
+                                                              "/etc/rc1.d",
+                                                              "/etc/rc2.d",
+                                                              "/etc/rc3.d",
+                                                              "/etc/rc4.d",
+                                                              "/etc/rc5.d",
+                                                              "/etc/rc6.d",
+                                                              "/etc/rcS.d"};
+
+  std::unordered_map<std::string, std::set<size_t>> service_list;
+
+  for (size_t run_level = 0U; run_level < run_level_dir_list.size();
+       ++run_level) {
+    const auto& run_level_path = run_level_dir_list[run_level];
+
+    try {
+      boostfs::directory_iterator end_it;
+      for (auto it = boostfs::directory_iterator(run_level_path); it != end_it;
+           ++it) {
+        const auto& symlink = it->path();
+        if (!boostfs::is_symlink(symlink)) {
+          continue;
+        }
+
+        auto symlink_destination = boostfs::canonical(symlink);
+        if (!boostfs::is_regular_file(symlink_destination)) {
+          VLOG(1) << "Skipping broken symlink: " << symlink << " -> "
+                  << symlink_destination;
+
+          continue;
+        }
+
+        auto service_name = symlink_destination.filename().string();
+        service_list[service_name].insert(run_level);
+      }
+    } catch (const boostfs::filesystem_error& e) {
+      VLOG(1) << "An error has occurred: " << e.what() << ". Continuing anyway";
+    }
+  }
+
+  for (auto it = service_list.begin(); it != service_list.end(); it++) {
+    const auto& service_name = it->first;
+    const auto& run_level_list = it->second;
+
+    Row r;
+    r["name"] = service_name;
+    r["status"] = "";
+    r["path"] = std::string("/etc/init.d/") + service_name;
+    r["service_type"] = "SysVinit";
+
+    // Use the same syntax as Upstart
+    if (!run_level_list.empty()) {
+      r["start_type"] = "runlevel [";
+
+      for (auto run_level : run_level_list) {
+        if (run_level != 7) {
+          r["start_type"] += std::to_string(run_level);
+        } else {
+          r["start_type"] += "S";
+        }
+      }
+
+      r["start_type"] += "]";
```

Comment:
Is it the best way to store it? It's not really queriable on something that way. Maybe others would have the opinion on that.

Reply:
I agree it is hard to query, but the table is supporting three different init systems (sysVinit, upstart, systemd). In the worst scenario, you have complex logic in the form of a script. There is really no easy way to flatten this information, and reporting the actual configuration content may at least give a chance to the administrator to access the correct information

Reply:
What about storing the bitmask? Will it be a bit easier to query things? @obelisk

Reply:
If you are talking about sysVinit services, this format has been chosen because it matches the one used by Upstart:

```
# initctl show-config cups
cups
  start on (filesystem and (started dbus or runlevel [2345]))
  stop on runlevel [016]
```